script/tool/lib/src/pubspec_check_command.dart

```diff
@@ -553,6 +554,23 @@ class PubspecCheckCommand extends PackageLoopingCommand {
         }
       });
     }
+
+    // Ensure that dev-only dependencies aren't in `dependencies`.
+    const List<String> devOnlyDependencies = <String>['integration_test'];
+    // pigeon/platform_tests/shared_test_plugin_code is allowed to violate
+    // the dev only dependencies rule beause pidgeon has generated tests that
+    // are intended to ship to customers.
+    if (pubspec.name == 'shared_test_plugin_code') {
```

Comment:
@stuartmorgan requesting re-review because I am not a fan of how I exempted pidgeon from the new check.

Reply:
You can just conditionalize this whole check on being publishable, and remove the warning:
https://github.com/flutter/packages/blob/dcb4f853b2476d5106aabe88186b1fad8ad623b4/script/tool/lib/src/pubspec_check_command.dart#L209